### PR TITLE
ip/ffmpeg.c: fix ffmpeg API usage

### LINKED ISSUE #812: APE playback freezes cmus
---

1) `uname -a`:
```
Linux _QzWxEc_-717 4.17.9-1-zen #1 ZEN SMP PREEMPT Sun Jul 22 20:23:46 UTC 2018 x86_64 GNU/Linux
```

2) `cmus --version`:
```
cmus v2.8.0-rc0-90-g17bf542
```

3) `cmus --plugins`:
```
Input Plugins: /usr/lib/cmus/ip
  mad:
    Priority: 55
    File Types: mp3 mp2
    MIME Types: audio/mpeg audio/x-mp3 audio/x-mpeg
  aac:
    Priority: 50
    File Types: aac
    MIME Types: audio/aac audio/aacp
  mpc:
    Priority: 50
    File Types: mpc mpp mp+
    MIME Types: audio/x-musepack
  opus:
    Priority: 50
    File Types: opus
    MIME Types:
  flac:
    Priority: 50
    File Types: flac fla
    MIME Types:
  modplug:
    Priority: 50
    File Types: mod s3m xm it 669 amf ams dbm dmf dsm far mdl med mtm okt ptm stm ult umx mt2 psm
    MIME Types:
  wavpack:
    Priority: 50
    File Types: wv
    MIME Types: audio/x-wavpack
  mp4:
    Priority: 50
    File Types: mp4 m4a m4b
    MIME Types:
  vorbis:
    Priority: 50
    File Types: ogg oga ogx
    MIME Types: application/ogg audio/x-ogg
  cue:
    Priority: 50
    File Types:
    MIME Types: application/x-cue
  cdio:
    Priority: 50
    File Types:
    MIME Types: x-content/audio-cdda
  wav:
    Priority: 50
    File Types: wav
    MIME Types:
  ffmpeg:
    Priority: 30
    File Types: aa aac ac3 aif aifc aiff ape au fla flac m4a m4b mka mkv mp+ mp2 mp3 mp4 mpc mpp ogg shn tak tta wav webm wma wv
    MIME Types:

Output Plugins: /usr/lib/cmus/op
  pulse
  alsa
  oss
  jack
  ao
```

After the first track finishes from an APE file in the playlist, cmus locks up. FLAC playback works perfectly fine. This only happens with APE files. The APE cuesheet is being used to create the tracklist and play via the "Library view (1)".

ffmpeg 1:4.0.2-1 on Arch Linux

### LINKED ISSUE #895: WMA freezes cmus
cmus freezes when playing wma files. SIGKILL is required to stop cmus (SIGTERM doesn't work). ffplay can play the files correctly.

My best guess is that the issue is with the ffmpeg plugin. I used to be able to play wma without issue, and I recently updated ffmpeg.

Here's the backtrace from gdb
```
#0  0x00007ffff7bb8b48 in __memmove_sse2_unaligned_erms () from /usr/lib/libc.so.6 
#1  0x00007ffff494147b in avpriv_copy_bits () from /usr/lib/libavcodec.so.58   
#2  0x00007ffff4fcad59 in ?? () from /usr/lib/libavcodec.so.58 
#3  0x00007ffff4fcfdbd in ?? () from /usr/lib/libavcodec.so.58   
#4  0x00007ffff49daa60 in ?? () from /usr/lib/libavcodec.so.58 
#5  0x00007ffff49db768 in avcodec_send_packet () from /usr/lib/libavcodec.so.58   
#6  0x00007ffff6823ab5 in ffmpeg_fill_buffer (ic=0x5555558cb4c0, cc=0x5555558961c0, input=0x5555558c8520,  
    output=0x5555558c7090, swr=0x5555559fdac0) at ip/ffmpeg.c:338   
#7  0x00007ffff6823c85 in ffmpeg_read (ip_data=0x555555888538,     
    buffer=0x55555567e6b0 "Q(\035\032a'C\031\220&m\030\342&\214\030\017'{\030J&9\027e%\336\025\023%\254\025]%M\026\064%D\026\005$\357\024\027#v\023\320\"\250\022&\"", count=60480) at ip/ffmpeg.c:394 
#8  0x000055555558112b in ip_read (ip=0x555555888530,
    buffer=0x55555567e6b0 "Q(\035\032a'C\031\220&m\030\342&\214\030\017'{\030J&9\027e%\336\025\023%\254\025]%M\026\064%D\026\005$\357\024\027#v\023\320\"\250\022&\"", count=60480) at input.c:695 
#9  0x000055555558b4d0 in _prebuffer () at player.c:599
#10 0x000055555558cc05 in player_set_file (ti=0x7fffcc006180) at player.c:1171    
#11 0x0000555555574136 in run_command (buf=<optimized out>) at command_mode.c:2883
#12 0x000055555556a7a1 in u_getch () at ui_curses.c:2163   
#13 main_loop () at ui_curses.c:2272test-song-6.mp2 
#14 main (argc=<optimized out>, argv=<optimized out>) at ui_curses.c:2556     
```

1) `uname -a`:
```
Linux hostname 5.0.5-arch1-1-ARCH #1 SMP PREEMPT Wed Mar 27 17:53:10 UTC 2019 x86_64 GNU/Linux
```

2) `cmus --version`:
```
cmus v2.8.0-16-ge429ddf
```

3) `cmus --plugins`:
```
Input Plugins: /usr/lib/cmus/ip
  mad:
    Priority: 55
    File Types: mp3 mp2
    MIME Types: audio/mpeg audio/x-mp3 audio/x-mpeg
  flac:
    Priority: 50
    File Types: flac fla
    MIME Types:
  aac:
    Priority: 50
    File Types: aac
    MIME Types: audio/aac audio/aacp
  wavpack:
    Priority: 50
    File Types: wv
    MIME Types: audio/x-wavpack
  mpc:
    Priority: 50
    File Types: mpc mpp mp+
    MIME Types: audio/x-musepack
  wav:
    Priority: 50
    File Types: wav
    MIME Types:
  modplug:
    Priority: 50
    File Types: mod s3m xm it 669 amf ams dbm dmf dsm far mdl med mtm okt ptm stm ult umx mt2 psm
    MIME Types:
  cue:
    Priority: 50
    File Types:
    MIME Types: application/x-cue
  opus:
    Priority: 50
    File Types: opus
    MIME Types:
  cdio:
    Priority: 50
    File Types:
    MIME Types: x-content/audio-cdda
  vorbis:
    Priority: 50
    File Types: ogg oga ogx
    MIME Types: application/ogg audio/x-ogg
  mp4:
    Priority: 50
    File Types: mp4 m4a m4b
    MIME Types:
  ffmpeg:
    Priority: 30
    File Types: aa aac ac3 aif aifc aiff ape au fla flac m4a m4b mka mkv mp+ mp2 mp3 mp4 mpc mpp ogg opus shn tak tta wav webm wma wv
    MIME Types:

Output Plugins: /usr/lib/cmus/op
  pulse
  alsa
  oss
  jack
```

### LINKED ISSUE #1063: 99% CPU at *end* of song
Sometimes cmus freezes and gets locked into around 98-100% CPU usage, and can only be killed by a SIGKILL. I've discovered for me, unlike other issues that experience this, it happens at the _end_ of a song. It freezes once the song finishes and is _about_ to move to the next one. But when I play the next one without finishing the previous, it works fine.

Some conveniently free examples of music that triggers this for me (fan-made rearrangements of videogame music, because I'm _that_ person :stuck_out_tongue: ):

- https://ocremix.org/remix/OCR03483 "Dream of Green"
- https://ocremix.org/remix/OCR03019 "A Derelict Machine"

Most of the music there is fine. I haven't found a pattern.

---

1) `uname -a`:
```
Linux changeling 5.3.18-lp152.66-default #1 SMP Tue Mar 2 13:18:19 UTC 2021 (73933a3) x86_64 x86_64 x86_64 GNU/Linux
```

2) `cmus --version`:
```
cmus v2.9.1
Copyright 2004-2006 Timo Hirvonen
Copyright 2008-2016 Various Authors
```

3) `cmus --plugins`:
```
Input Plugins: /usr/lib64/cmus/ip
  wav:
    Priority: 50
    File Types: wav
    MIME Types:
  cue:
    Priority: 50
    File Types:
    MIME Types: application/x-cue
  flac:
    Priority: 50
    File Types: flac fla
    MIME Types:
  ffmpeg:
    Priority: 30
    File Types: aa aac ac3 aif aifc aiff ape au fla flac m4a m4b mka mkv mp+ mp2 mp3 mp4 mpc mpp ogg opus shn tak tta wav webm wma wv *
    MIME Types:

Output Plugins: /usr/lib64/cmus/op
  pulse
  alsa
  oss
```

### LINKED ISSUE #1081: cmus does not play an m4a file, instead using 100% CPU and requiring a SIGKILL
See title.

cmus uses the ffmpeg plugin to open the m4a.

There are existing issues related to CPU use and m4a files (separately). I'm not sure if they're related or not.

Files: https://gist.github.com/szc126/84da78d242f50cc14077a0ec63833ba8

---

1) `uname -a`:
```
Linux myhostname 5.12.1-arch1-1 #1 SMP PREEMPT Sun, 02 May 2021 12:43:58 +0000 x86_64 GNU/Linux
```

2) `cmus --version`:
```
cmus v2.9.1
Copyright 2004-2006 Timo Hirvonen
Copyright 2008-2016 Various Authors
```

3) `cmus --plugins`:
```
Input Plugins: /usr/lib/cmus/ip
  mad:
    Priority: 55
    File Types: mp3 mp2
    MIME Types: audio/mpeg audio/x-mp3 audio/x-mpeg
  vorbis:
    Priority: 50
    File Types: ogg oga ogx
    MIME Types: application/ogg audio/x-ogg
  cue:
    Priority: 50
    File Types:
    MIME Types: application/x-cue
  wav:
    Priority: 50
    File Types: wav
    MIME Types:
  cdio:
    Priority: 50
    File Types:
    MIME Types: x-content/audio-cdda
  flac:
    Priority: 50
    File Types: flac fla
    MIME Types:
  modplug:
    Priority: 50
    File Types: mod s3m xm it 669 amf ams dbm dmf dsm far mdl med mtm okt ptm stm ult umx mt2 psm
    MIME Types:
  aac:
    Priority: 50
    File Types: aac
    MIME Types: audio/aac audio/aacp
  wavpack:
    Priority: 50
    File Types: wv
    MIME Types: audio/x-wavpack
  ffmpeg:
    Priority: 30
    File Types: aa aac ac3 aif aifc aiff ape au fla flac m4a m4b mka mkv mp+ mp2 mp3 mp4 mpc mpp ogg opus shn tak tta wav webm wma wv
    MIME Types:

Output Plugins: /usr/lib/cmus/op
  pulse
  alsa
  oss
  jack
```

### PR DESCRIPTION
**Pull request by dryleev**
Fixes #812
Fixes #895
Fixes #1063
Fixes #1081

Credits: @gdrbyKo1

### CODE REVIEW

**Review by pgaskin — APPROVED** [review]
Thanks! I've tested this, and it works well. I've also did some research about `avcodec_send_packet` to confirm.

 @nefthy?

**Comment by pgaskin** [comment]
Also, @flyingmutant, you'll need to approve the workflow runs for new PRs now.

**Comment by flyingmutant** [comment]
Thanks! Merged.

File: ip/ffmpeg.c [hunk 1]
```diff
@@ -335,16 +335,27 @@ static int ffmpeg_fill_buffer(AVFormatContext *ic, AVCodecContext *cc, struct ff
...
-				got_frame = !avcodec_receive_frame(cc, frame);
-				if (got_frame)
-					len = input->curr_pkt_size;
-				else
-					len = 0;
-			} else {
-				got_frame = 0;
+			int send_result = avcodec_send_packet(cc, &avpkt);
+			if (send_result != 0) {
+				if (send_result != AVERROR(EAGAIN)) {
```
[the remark this answers is not in the document]
Comment (dryleev) [reply ip/ffmpeg.c:340]: Only by setting `len` to `0` in order to bypass input buffer consumption. Maybe would be better to initialize `len` with `0` where it is [declared](https://github.com/dryleev/cmus/blob/8e40af2379f94f5e18fe7c3f79214c3056220d52/ip/ffmpeg.c#L307) and get rid of line [352](https://github.com/dryleev/cmus/blob/8e40af2379f94f5e18fe7c3f79214c3056220d52/ip/ffmpeg.c#L352).

[the remark this answers is not in the document]
Comment (pgaskin) [reply ip/ffmpeg.c:340]: The advantage of keeping it the way it is right now is that code analysis tools will warn if len isn't set by the other code paths if it's left undeclared.

**Pull request merged**